src/libs/elektra/hooks.c

```diff
@@ -81,6 +101,64 @@ static int initHooksSpec (KDB * kdb, Plugin * plugin, Key * errorKey)
 	return 0;
 }
 
+static int initHooksSendNotifications (KDB * kdb, const KeySet * config, KeySet * modules, const KeySet * contract, Key * errorKey)
+{
+	struct _SendNotificationHook * lastHook = kdb->hooks.sendNotification = NULL;
+
+	Key * pluginsKey = keyNew ("system:/elektra/hook/notification/send/plugins", KEY_END);
+	KeySet * configuredPlugins = elektraArrayGet (pluginsKey, config);
+
+	if (ksGetSize (configuredPlugins) == 0)
+	{
+		return 0;
+	}
+
+	ksRewind (configuredPlugins);
+
+	Key * cur;
+	while ((cur = ksNext(configuredPlugins)) != NULL)
+	{
+		const char * pluginName = keyString (cur);
+		Plugin * plugin = loadPlugin (pluginName, kdb->global, modules, contract, errorKey);
+
+		if (!plugin)
+		{
+			continue;
+		}
+
+		kdbHookSendNotificationGetPtr getPtr = (kdbHookSendNotificationGetPtr) getFunction (plugin, "hook/notification/send/get", errorKey);
+		kdbHookSendNotificationSetPtr setPtr = (kdbHookSendNotificationSetPtr) getFunction (plugin, "hook/notification/send/set", errorKey);
+
+		if (!getPtr || !setPtr)
+		{
+			elektraPluginClose (plugin, errorKey);
+			continue;
+		}
```

Comment:
Should we really require both hooks? From a functionality standpoint it doesn't matter to `libelektra-kdb`. We don't expect to get any result from a notification plugin, so unlike with `spec` it's not necessarily wrong to just provide one of the hooks.

For example, there could be an implementation that just uses `hook/notification/send/set` and basically just broadcasts "hey someone wrote to the KDB". Or a very simple auditing mechanism that just records every single `kdbSet` call.

I'm not sure there is an example of just using `hook/notification/send/get`. But as I said, `libelektra-kdb` doesn't expect any particular outcomes so we don't really care which hooks are available.

Reply:
Yes, I agree. I made both optional, and issue a warning when none of them is set.